micro-rdk-nmea/src/messages/pgns.rs

```diff
@@ -106,3 +115,96 @@ pub struct GnssSatsInView {
     #[length_field = "sats_in_view"]
     satellites: Vec<Satellite>,
 }
+
+macro_rules! define_pgns {
+    ( $(($pgndef:ident, $enum:ident, $pgn:expr)),* ) => {
+        #[derive(Clone, Debug)]
+        pub enum MessageData {
+            $($enum($pgndef)),*,
+            Unsupported(UnparsedMessageData)
+        }
+
+        impl MessageData {
+            pub fn pgn(&self) -> u32 {
+                match self {
+                    $(Self::$enum(_) => $pgn),*,
+                    Self::Unsupported(unparsed) => unparsed.pgn()
+                }
+            }
+
+            pub fn from_bytes(pgn: u32, bytes: Vec<u8>) -> Result<Self, crate::parse_helpers::errors::NmeaParseError> {
+                Ok(match pgn {
+                    $($pgn => {
+                        let cursor = DataCursor::new(bytes);
+                        Self::$enum($pgndef::from_cursor(cursor)?)
```

Comment:
If each struct as an associated PGN, is there a double check you can do here that the runtime value of `pgn` matches the one statically associated with the struct?

Reply:
No because the runtime value of the PGN is what determines which struct will be used (see `NmeaMessage::new`)

Reply:
I meant it as more sort of a debug assertion. Once you have the type, the statically associated ID on the type better be equal to the runtime type or somethign went terribly wrong. But I'm mostly OK with your argument that that can't happen.